src/backend/distributed/master/worker_node_manager.c

```diff
@@ -355,6 +356,73 @@ ActivePrimaryNodeList(LOCKMODE lockMode)
 }
 
 
+/*
+ * ActivePrimaryDesiredDataNodeList returns a list of all active, primary
+ * worker nodes that can store new data, i.e isdatanode is 'true'.
+ */
+List *
+ActivePrimaryDesiredDataNodeList(LOCKMODE lockMode)
```

Comment:
This function and the function below share 99% if the code except for a function call to either `WorkerNodeIsPrimaryDesiredDataNode` and `WorkerNodeIsPrimaryCurrentDataNode`.

Any way we can share more of the code instead of copying?